docs/types.md

```diff
@@ -0,0 +1,276 @@
+
+Once you understand the problem well enough to have the data structures
+laid out, the algorithms usually fall into place.
+
+I've been programming for over 35 years, and I still experience a bit
+of shock and joy when a program works perfectly the first time I run
+it.
+
+## Type Inference
+
+The compiler will examine a program and try to solve all the types in
+the program, using the
+[Hindley-Milner](https://en.wikipedia.org/wiki/Hindley%E2%80%93Milner_type_system)
+algorithm.
+
+One of the goals of Irken is to reduce the use of type annotations.  I
+believe they are one of the major turn-offs of functional languages,
+especially for newcomers.  So I want Irken to _look_ like Scheme (or
+many other high level languages), but be something completely
+different underneath.
+
+Irken supports type annotations, and in some cases they are necessary
+for type safety, but for the most part they can be left off.
+
+Note: the compiler is capable of dumping all the types in a program if
+you need them for documentation or understanding. (give the compiler
+the `-types` option).
+
+### A simple example.
+
+Here's a factorial function:
+
+```scheme
+(define (fact n)
+  (if (= n 0)
+      1
+      (* n (fact (- n 1)))
+      ))
+```
+
+When the compiler walks this function, it begins with the innermost
+expressions.  For example, the `(= n 0)` expression contains lots of
+usable type information already.  We know that `0` is an `int`.  We
+know that `=` takes two `int`s as arguments and returns a `bool`.  We
+combine this information to assert that `n` is also an integer.
+
+The `if` statement requires a `bool` as its first argument.  So that part
+checks out - our test returns a `bool`.
+
+The 'then' clause of the `if` returns `1`, which is an `int`.  So now
+we know that the value returned by the `if` statement must be an
+`int`.  That tells us that the function itself returns an `int`.  At
+this point, we know the full type signature for the function: it takes
+an `int` argument and returns an `int`.
+
+The 'else' clause follows similarly.
+
+### The generated code.
+
+What does this mean for the generated code?  It means that it never
+needs to check the type of anything it touches.  When the `if` becomes
+a machine-level branch, it does an exact integer comparison to the
+known value of `#f` to decide wether to take the branch:
+
+Here's what the C backend produces for that `if`:
+
+```C
+
+  O r60 = (object *) 1; // representation of '0'
+  O r61 = ((object*) lenv) [4]; // fetch of 'n'
+  if IRK_IS_TRUE(IRK_TEST(UNTAG_INTEGER(r60)<=UNTAG_INTEGER(r61))) { // (= n 0)
+    ...
+  }
+```
+
+After optimizing by the C compiler, that test turns into:
+
+```asm
+        movq    _lenv(%rip), %rax # load 'n'
+        cmpq    $1, 16(%rax)      # compare to '0'
+        ja      LBB54_4           # branch
+```
+
+## The Type System.
+
+Irken uses a small set of 'base types':
+
+  * integers (signed, 63-bit on a 64-bit machine)
+  * booleans
```

Comment:
does irken follow the convention of treating "non-empty" instances as True and empty ones as False? I'm going to guess "no"